src/test/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/AlkylStructureFragmenterTest.java

```diff
@@ -279,19 +288,27 @@ private boolean compareListsIgnoringOrder(ArrayList aList1, ArrayList aList2) {
     /**
      * Utility method generating SMILES notation strings for a given AtomContainerSet.
      *
-     * @param anAtomContainer given AtomContainerSet
+     * @param anACSet given AtomContainerSet
      * @return List containing the generated Strings
      * @throws CDKException if SmilesGenerator is unable to generate String from structure
      */
-    private List<String> generateSMILESFromACSet(IAtomContainer anAtomContainer) throws CDKException {
+    private List<String> generateSMILESFromACSet(IAtomContainerSet anACSet) throws CDKException {
         SmilesGenerator tmpSmilesGenerator = new SmilesGenerator(SmiFlavor.Canonical);
-        List<String> tmpSmilesList = new ArrayList<>(anAtomContainer.getAtomCount());
-        tmpSmilesList.add(tmpSmilesGenerator.create(anAtomContainer));
+        List<String> tmpSmilesList = new ArrayList<>(anACSet.getAtomContainerCount());
+        CDKHydrogenAdder tmpAdder = CDKHydrogenAdder.getInstance(new SilentChemObjectBuilder());
+        for (IAtomContainer tmpAC : anACSet.atomContainers()) {
+            AtomContainerManipulator.percieveAtomTypesAndConfigureAtoms(tmpAC);
+            tmpAdder.addImplicitHydrogens(tmpAC);
+            Kekulization.kekulize(tmpAC);
```

Comment:
It seems the kekulization of the resulting fragments after successful fragmentation leads to an exception originating from unset implicit hydrogens, even though they should be added by the CDKHydrogenAdder before starting kekulization.

Reply:
Apparently, you have two fragments of SMILES code "*C(*)*" in your resulting fragments set. I have looked at the detected atom types and their implicit hydrogen counts. Somehow, I don't know how, they are treated differently here. In the first structure, every R-/pseudo-atom has a hydrogen count of 0, and in the second structure, it is "null", causing the issue.But the CDKHydrogenAdder should ignore all of the pseudo atoms (see here: https://github.com/cdk/cdk/blob/01d7ab0ea4f5f0f590b42cbb2c45e32e2cf41066/base/valencycheck/src/main/java/org/openscience/cdk/tools/CDKHydrogenAdder.java#L98). An interesting bug or feature: if you use the atom-specific method, it does not check for pseudo-atoms and treats them "right", i.e. assigns an implicit hydrogen count of 0 to them. Then, your test does not throw an exception but the assertions fails.

Reply:
>I have looked at the detected atom types and their implicit hydrogen counts. Somehow, I don't know how, they are treated differently here. In the first structure, every R-/pseudo-atom has a hydrogen count of 0, and in the second structure, it is "null", causing the issue.

I saw that too and was very confused by it.
>An interesting bug or feature: if you use the atom-specific method, it does not check for pseudo-atoms and treats them "right", i.e. assigns an implicit hydrogen count of 0 to them.

I will test this too in the coming days.
>We can discuss the theoretical background on Friday.

Gladly, I would like to know the underlying theory.